### PR TITLE
Macos mode remove_cert

### PR DESCRIPTION
**Pull request by emanuele-em**
## Summary

* Added the ability to delete the previously added and trusted mitmproxy certificate:
* Utilized the [security-framework crate](https://docs.rs/security-framework/latest/security_framework/certificate/struct.SecCertificate.html#method.delete) `delete` method from [SecItemDelete](https://developer.apple.com/documentation/security/1395547-secitemdelete).
* The deletion is based on the **label** parameter, [as recommmended](https://developer.apple.com/documentation/security/ksecattrlabel) by the security framework documentation. This **label** was initially set when the certificate was added.

## Checklist
- [x] Applied `cargo fmt`
- [x] No error nor warning in the console after running `maturin develop`

### CODE REVIEW

File: ffi/src/lib.rs [hunk 1]
```diff
@@ -42,6 +42,8 @@ pub fn mitmproxy_rs(_py: Python, m: &PyModule) -> PyResult<()> {
     m.add_class::<server::WireGuardServer>()?;
     m.add_function(wrap_pyfunction!(util::genkey, m)?)?;
     m.add_function(wrap_pyfunction!(util::pubkey, m)?)?;
+    m.add_function(wrap_pyfunction!(util::add_trusted_cert, m)?)?;
+    m.add_function(wrap_pyfunction!(util::delete_cert, m)?)?;
```
Comment (mhils) [inline ffi/src/lib.rs:46]: `remove_trusted_cert`? (to have some symmetry in the API)

File: ffi/src/util.rs [hunk 2]
```diff
@@ -1,4 +1,5 @@
 use data_encoding::BASE64;
+use mitmproxy::macos;
```
Comment (mhils) [inline ffi/src/util.rs:2]: This needs a `#[cfg(...)]`, otherwise CI is failing. :)

File: ffi/src/util.rs [hunk 3]
```diff
@@ -61,3 +62,41 @@ pub fn pubkey(private_key: String) -> PyResult<String> {
...
+    {
+        let pem_body = pem
+            .lines()
+            .skip(1)
+            .take_while(|&line| line != "-----END CERTIFICATE-----")
+            .collect::<String>();
+        let der = BASE64.decode(pem_body.as_bytes()).unwrap();
+        match macos::add_trusted_cert(der) {
+            Ok(_) => Ok(()),
+            Err(_) => Err(PyErr::new::<PyOSError, _>("Invalid certificate")),
```
Comment (mhils) [inline ffi/src/util.rs:79]:
```suggestion
            Err(e) => Err(PyErr::new::<PyOSError, _>(format!("Failed to install certificate: {:?}", e))),
```

Would something like this be useful to aid debugging bug reports from users? :)

File: macos-trust-cert.app/Contents/Info.plist [hunk 4]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple Computer//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+  <key>CFBundleDisplayName</key>
+  <string>macos-trust-cert</string>
+  <key>CFBundleExecutable</key>
+  <string>add_certificate_test</string>
+  <key>CFBundleIdentifier</key>
+  <string>com.mitmproxy.mitmproxy</string>
```
Comment (mhils) [inline macos-trust-cert.app/Contents/Info.plist:10]: FWIW let's use `org.mitmproxy.cert_install` here, we don't really use mitmproxy.com.

File: macos-trust-cert.app/Contents/Info.plist [hunk 5]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple Computer//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+  <key>CFBundleDisplayName</key>
+  <string>macos-trust-cert</string>
```
Comment (mhils) [inline macos-trust-cert.app/Contents/Info.plist:6]: Is this surfaced in any UI somewhere? If yes, probably makes sense to identify it as mitmproxy here, e.g. `mitmproxy certificate installer`. If that's not the case it does not matter obviously. :)

File: macos-trust-cert.app/Contents/MacOS/add_certificate_test
Comment (mhils) [inline macos-trust-cert.app/Contents/MacOS/add_certificate_test:1]: Can we easily bring the sources for this into the repo, or is that tricky currently due to modified dependencies? :)

File: macos-trust-cert.app/Contents/Resources/_root_/Users/emanuelemicheletti/.mitmproxy/mitmproxy-ca-cert.pem [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+X2IupwP7LzRof5ZsYE6YY5hwTcITnzMCAwEAAaNXMFUwDwYDVR0TAQH/BAUwAwEB
+/zATBgNVHSUEDDAKBggrBgEFBQcDATAOBgNVHQ8BAf8EBAMCAQYwHQYDVR0OBBYE
+FILZ0JOvphBddkLzuYWinADisft4MA0GCSqGSIb3DQEBCwUAA4IBAQCEnQdbZrdO
+UZ/b6cxmYRGgl1vUkW++zkaFyCyy6/78twoEvrTjkYpkuAqYYyWY3q+KlpyWPKSo
+MRdu/a6tdk6Ueuhdlw0k/az92syLRAHtvApe/2DmaBfuwRcygX4gM0Xza40tkp1I
+DaYagu4SPfYKeBfKXxZ/0aYEgrLY+k1BcMjXqD0YldzMaWklK3p7rBYcT0b1vdAU
+F9/P6cUSQpwEFQrXItYyxDLOXcOJfllCyZkFd3f6HhuUd2EyVrfGyGQZlFTEuFvE
+35dwgGTmbpgHjCSP/VR7qD03FIkXisjJBwnQBSxRiQ5MiQ36hO60cEYtHL0oxUdj
+Wfo3v7rwNfhH
+-----END CERTIFICATE-----
```
Comment (mhils) [inline macos-trust-cert.app/Contents/Resources/_root_/Users/emanuelemicheletti/.mitmproxy/mitmproxy-ca-cert.pem:20]: Added intentionally?

File: src/macos.rs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    let cert = SecCertificate::from_der(&der)?;
+    let add_ref = AddRef::Certificate(cert.clone());
+    let add_option = ItemAddOptions::new(ItemAddValue::Ref(add_ref))
+        .set_label("mitmproxy")
+        .to_dictionary();
+
+    ItemSearchOptions::new()
+        .class(ItemClass::certificate())
+        .load_refs(true)
+        .label("mitmproxy")
```
Comment (mhils) [inline src/macos.rs:20]: What happens if there are multiple certificates named "mitmproxy"?

File: src/macos.rs [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        .search()
+        .and_then(|_| Ok(()))
+        .unwrap_or_else(|_| add_item(add_option).unwrap());
+
+    if let Err(err) = std::process::Command::new("open")
+        .arg("../macos-trust-cert.app")
+        .spawn()
+        .unwrap()
```
Comment (mhils) [inline src/macos.rs:28]: We shouldn't panic here, but maybe return an error? :)

File: src/macos.rs [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+    ItemSearchOptions::new()
+        .class(ItemClass::certificate())
+        .load_refs(true)
+        .label("mitmproxy")
+        .search()
+        .and_then(|_| Ok(()))
+        .unwrap_or_else(|_| add_item(add_option).unwrap());
+
+    if let Err(err) = std::process::Command::new("open")
```
Comment (mhils) [inline src/macos.rs:25]: Is this blocking the mitmproxy process entirely? If it is, maybe use https://docs.rs/tokio/latest/tokio/process/? :)

File: src/macos.rs [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    Ok(())
+}
+
+pub fn delete_cert() -> Result<()> {
+    if let SearchResult::Ref(Reference::Certificate(cert)) = ItemSearchOptions::new()
+        .class(ItemClass::certificate())
+        .load_refs(true)
+        .label("mitmproxy")
+        .search()
+        .unwrap()
```
Comment (mhils) [inline src/macos.rs:42]: This method should also never panic but return errors instead. We're already using `anyhow` a bit, which makes that relatively painless. :)

**Review by mhils — COMMENTED** [review]
Nice progress! 🚀

File: macos-trust-cert.app/Contents/MacOS/add_certificate_test
Comment (emanuele-em) [reply macos-trust-cert.app/Contents/MacOS/add_certificate_test:1]: the file is compiled, so since I compiled it with the dependencies already modified I don't think it creates any problems, it should be completely unrelated to the external sources, or should it? (otherwise it would not be portable I think)

Comment (mhils) [reply macos-trust-cert.app/Contents/MacOS/add_certificate_test:1]: Not sure if I understand your comment. Can you rephrase? :)

Comment (emanuele-em) [reply macos-trust-cert.app/Contents/MacOS/add_certificate_test:1]: I may be saying something **stupid**.
you were referring to the rust-security-framework right? In this case, since that `add_certificate_test` is compiled, it should not be a problem

Comment (mhils) [reply macos-trust-cert.app/Contents/MacOS/add_certificate_test:1]: > I may be saying something **stupid**.

Not at all. :-)

Let me rephrase my initial question: If I understand things correctly, `add_certificate_test` is a Rust binary you've written that adds the CA certificate to the keychain. It depends on a modified fork of `rust-security-framework` (that currently only exists on your local disk). Is that correct so far?

If my interpretation above is correct, if you would [get hit by a bus](https://en.wikipedia.org/wiki/Bus_factor) tomorrow, I think I'd have no chance of ever updating that binary other than rewriting it from scratch. Long story short, it would be incredibly useful to have the binary source code in the repository. You can take a look at `wireguard-test-client` in the repo on what this would look like. For `rust-security-framework`, you can push your modifications to GitHub and then reference your repo in `Cargo.toml`. :)

Let me know if that clears things up!
